### DISCUSSION TITLE
Another PERMIT2 Mystery

### DISCUSSION THREAD

**Discussion by takabinance** [Elnaril/uniswap-universal-router-decoder#23] [Q&A]
Have been struggling to figure out how to get permit2 to give allowance to universal router. I have confirmed that permit2 has a max allowance to the token (WETH) I am trying to swap. When I execute the following code (running on Metis L2, Unimaia V3), the contract always reverts. 

If I manually approve WETH in the UI, then everything else works fine (e.g., swaps). But am hoping it's just some simple thing I'm overlooking. I've also tried combining this with a v3 swap and has same error. 

Things of note:
the permit2_contract.functions.allowance() returns [0, 0, 0]

Here is a failed tx that I submitted:
https://andromeda-explorer.metis.io/tx/0x1d8d7793f0c0e515dcb9b32eb2f56bd8d58674422964b3828c019c0915009fc1

Any ideas?

```
from eth_defi.gas import estimate_gas_fees, apply_gas
from uniswap_universal_router_decoder import RouterCodec
from web3._utils.transactions import fill_nonce
from web3 import Account, Web3
from data.queries import *
from web3.middleware import geth_poa_middleware

w3 = Web3(Web3.HTTPProvider('https://andromeda.metis.io'))
w3.middleware_onion.inject(geth_poa_middleware, layer=0)
w3.eth.set_gas_price_strategy(lambda w3, transaction_params=None: w3.eth.gas_price)

permit2_addr = '0x2EE5407017B878774b58c34A8c09CAcC94aDd69B'
univ_router = '0x262E75e4DE2942181bD18DA2360be9699ebf051A'

permit2_abi = json.loads(open('permit2.json').read())
univ_router_abi = json.loads(open('universal_router.json').read())

permit2_contract = w3.eth.contract(abi=permit2_abi, address=permit2_addr)
univ_router_contract = w3.eth.contract(address=univ_router, abi=univ_router_abi)

pk = 'x'
account = Account.from_key(pk.replace(' ', ''))

from_token = '0x420000000000000000000000000000000000000A'  # WETH
to_token = '0xEA32A96608495e54156Ae48931A7c20f0dcc1a21'  # USDC
amount_in = 0.05 * 10 ** 18
fee_level = 3000
gas = 350000

codec = RouterCodec()
result = permit2_contract.functions.allowance(account.address, from_token, univ_router).call()
permit2_allowance, permit2_expiration, permit2_nonce = result[0], result[1], result[2]
print(permit2_allowance, permit2_expiration, permit2_nonce)   # All 0, 0, 0

deadline = expiration = int(time.time()) + 60 * 60
permit2_data, signable_message = codec.create_permit2_signable_message(
    from_token,
    2 ** 160 - 1,
    expiration,
    permit2_nonce,
    univ_router,
    deadline,
    1088,
)
signed_message = account.sign_message(signable_message)
# We now have the signed message with allowance for UR to spend from_token from permit2

encoded_data = codec.encode.chain().permit2_permit(permit2_data, signed_message).build(deadline)
tx = {
    'from': account.address,
    'to': univ_router,
    'data': encoded_data,
    'chainId': 1088,
    'gas': gas
}
tx = fill_nonce(w3, tx)
gas_fees = estimate_gas_fees(w3)
apply_gas(tx, gas_fees)

# Sign and send
signed_tx = account.sign_transaction(tx)
tx_hash = w3.eth.send_raw_transaction(signed_tx.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
print(tx_receipt)

```

**Comment by takabinance** [comment]
After approving manually (as part of a swap), when I come back to get permit2_contract.functions().allowance(), it correctly updates permit2_allowance, permit2_expiration and permit2_nonce (which is now 1). My script still fails.

**Comment by Elnaril** [comment]
Hello @takabinance !

That's a tough one indeed, thank you for sharing! 🤣

Your code looks good to me.
Moreover I was able to reproduce your issue on a local Metis fork using Ganache.

But the same code runs fine on a local Ethereum fork.
The only differences in the code are:

- The rpc endpoint used to access the blockchain
- The block number to fork from
- The weth, ur and permit2 addresses

I can't figure out what could be the Metis specificity that leads to this EVM Error.

If you, or anyone, find the root cause, please share it here as that tickles my curiosity! :)

**Comment by takabinance** [comment]
Thanks for looking at this. I'm going to try to dig into it today but have pretty basic evm skills. Is there any particular path you might recommend for me to find cause.

**Comment by Elnaril** [comment]
I'm not familiar with Metis, but maybe its community can give you a clue or two. Maybe someone got the same issue with the JS SDK.
It does not seem to be an official Uniswap router/permit2, so maybe you can reach out the dex behind it.

Thank you for your interest in this lib! :)

**Comment by Elnaril** [comment]
Hello @takabinance  !
The latest version (v1.2.0) should fix this issue. Just override the `verifying_contract` in `create_permit2_signable_message()` with the   actual Permit2 address on Metis.

**Comment by ronaldoguedess** [comment]
I'm looking for a Dex with support for Permit2, Pancakeswap removed it... 


Which one dex has this Permit2 support?
https://www.metis.io/ecosystem

this not give quotes:
https://v3.maiadao.io/#/swap

**Comment by Elnaril** [comment]
Hello @ronaldoguedess 
I'm not familiar enough with Metis to help you with its specificity, sorry ...
If not already done, I would recommend to ask this question to its community
